src/Temporalio/Worker/ActivityWorker.cs

```diff
@@ -0,0 +1,703 @@
+                WorkflowID: start.WorkflowExecution.WorkflowId,
+                WorkflowNamespace: start.WorkflowNamespace,
+                WorkflowRunID: start.WorkflowExecution.RunId,
+                WorkflowType: start.WorkflowType);
+            // Create context
+            var cancelTokenSource = new CancellationTokenSource();
+            var context = new ActivityContext(
+                info: info,
+                cancellationToken: cancelTokenSource.Token,
+                workerShutdownToken: workerShutdownTokenSource.Token,
+                tsk.TaskToken,
+                worker.Client.Options.LoggerFactory.CreateLogger(
+                    $"Temporalio.Activity:{info.ActivityType}"));
+
+            // Start task
+            using (context.Logger.BeginScope(info.LoggerScope))
+            {
+                // We know that the task for the running activity is accessed only on graceful
+                // shutdown which could never run until after this (and the primary execute) are
+                // done. So we don't have to worry about the dictionary having an activity without
+                // a task even though we're late-binding it here.
+                var act = new RunningActivity(context, cancelTokenSource);
+                runningActivities[tsk.TaskToken] = act;
+#pragma warning disable CA2008 // We don't have to pass a scheduler, factory already implies one
+                act.Task = worker.Options.ActivityTaskFactory.StartNew(
+                    () => ExecuteActivityAsync(act, tsk)).Unwrap();
+#pragma warning restore CA2008
+            }
+        }
+
+        private async Task ExecuteActivityAsync(
+            RunningActivity act,
+            Bridge.Api.ActivityTask.ActivityTask tsk)
+        {
+            // Try to build a completion, but if it fails then manually build one with the failure
+            Bridge.Api.ActivityTaskCompletion completion;
+            try
+            {
+                completion = await ExecuteActivityInternalAsync(act, tsk).ConfigureAwait(false);
+            }
+            catch (Exception e)
+            {
+                completion = new()
+                {
+                    Result = new()
+                    {
+                        Failed = new()
+                        {
+                            Failure_ = new() { Message = $"Failed building completion: {e}" },
+                        },
+                    },
+                };
+            }
+
+            // Try to finish heartbeats and send completion
+            try
+            {
+                // We have to wait on any outstanding heartbeats to finish. This will not throw. We
+                // accept that in a rare scenario, this heartbeat can fail to encode but it is too
+                // late to cancel the activity. Like other SDKs, we currently drop this heartbeat.
+                await act.FinishHeartbeatsAsync().ConfigureAwait(false);
```

Comment:
There's no need to wait for heartbeats to complete to send successful completion, only failure.

Reply:
And not for cancellation either.

Reply:
To be a proper steward of .NET and to properly ensure my Rust reference only has 1 remaining, I need to wait for all tasks to finish before finalizing shutdown of the worker.

Reply:
You're blocking completion from reaching the server which is more valuable than a heartbeat.

Reply:
It's valuable for tracking of async tasks and worker refs. This same thing exists for Python. We ensure every heartbeat called gets sent to Rust before completion so I don't have unaccounted for Rust work holding a ref to the worker. I know activity complete means no more Rust calls.

Reply:
Sure, but you can await this task after flushing the completion and prevent the call into core if the activity is already done.

Reply:
If you want to take it a step further (which I don't think is necessary), you can cancel the heartbeat task too.

Reply:
> Sure, but you can await this task after flushing the completion and prevent the call into core if the activity is already done.

Only in cases of success which complicates the code. I have added this complication (glad I didn't have to in Python). I question whether it makes a real difference (only in slow codecs on heartbeat + complete).